### PR TITLE
fix: dont NPE if initializing odo fails (#536)

### LINKED ISSUE #536: java.lang.NullPointerException Odo.getMasterUrl()
ENV: 
    Fedora 35

    Intellij IDEA 222.1 (Community Edition) bld #IC-222.3739.54
    Runtime version: 17.0.3+7-b469.37 amd64
    VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o

odo is in my classpath.  Verified by
  1668 > which odo
  ~/bin/odo

Exception in plugin OpenShift Toolkit by Red Hat (1.2.0.168)
java.lang.NullPointerException: Cannot invoke "org.jboss.tools.intellij.openshift.utils.odo.Odo.getMasterUrl()" because "this.delegate" is null
	at org.jboss.tools.intellij.openshift.utils.odo.OdoProjectDecorator.getMasterUrl(OdoProjectDecorator.java:258)
	at org.jboss.tools.intellij.openshift.tree.application.ApplicationsTreeStructure.lambda$createDescriptor$14(ApplicationsTreeStructure.java:246)
	at com.redhat.devtools.intellij.common.tree.LabelAndIconDescriptor.processLabel(LabelAndIconDescriptor.java:74)
	at com.redhat.devtools.intellij.common.tree.LabelAndIconDescriptor.update(LabelAndIconDescriptor.java:64)
	at com.intellij.ide.util.treeView.PresentableNodeDescriptor.getUpdatedPresentation(PresentableNodeDescriptor.java:78)
	at com.intellij.ide.util.treeView.PresentableNodeDescriptor.update(PresentableNodeDescriptor.java:30)
	at com.intellij.ui.tree.StructureTreeModel$Node.update(StructureTreeModel.java:458)
        ...

### PR DESCRIPTION
**Pull request by adietish**
fixes #536

### CODE REVIEW

**Comment by adietish** [comment]
@sbouchet this PR has 2 commits:
* 1st commit: simples fix
* 2nd commit: change in how odo is downloaded if it doesnt exist yet
Would love to get your take on this 2nd commit which I believe is an improvement/simplification of the whole approach

**Comment by sbouchet** [comment]
<snip>
> I tried to address this as follows:
> 
> 1. get rid of explicitly calling `initializeOdo()` . Load it transparently whenever `getOdo()`  is called
> 2. get rid of the instance var `odo` and `error`. Replace by an instance var `odoFuture` which is a `CompletableFuture`. It holds odo and the possible error.
> 
> As a result of this change any time user code wants to use the odo binary it has to try/catch to get errors that may have happened when loading the binary and null-check in case the odo binary wasnt loaded yet:
> 
> ```java
>  try {
>      Odo odo = componentNode.getRoot().getOdo().getNow(null);
>      if (odo == null) {
>          return;
>      }
>      ...
>  } catch (Exception ex) {
> ```

I like the idea to just call getOdo() that returns a future. this will also simplyfy the codebase by removing the odoprojectdecorator class. +1 for implement this.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java [hunk 1]
```diff
@@ -1,300 +0,0 @@
...
-    }
-
-    @Override
-    public List<ComponentMetadata> analyze(String path) throws IOException {
-        return delegate.analyze(path);
-    }
-
-    @Override
-    public void createComponent(String project, String componentType, String registryName, String component, String source, String devfile, String starter) throws IOException {
-        if (StringUtils.isNotBlank(starter)) {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java:82]: please move this logic in OdoCli or in the calling action

**Review by datho7561 — APPROVED** [review]
I can confirm that this presents the error in the OpenShift tree view when the tools.json is corrupted. The refactoring looks good as well. Thanks, Andre!

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java [hunk 2]
```diff
@@ -1,300 +0,0 @@
...
-    }
-
-    @Override
-    public void login(String url, String userName, char[] password, char[] token) throws IOException {
-        delegate.login(url, userName, password, token);
-    }
-
-    @Override
-    public List<Component> getComponents(String project) throws IOException {
-        List<Component> components = delegate.getComponents(project);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java:185]: please move this logic i in the calling action

File: src/main/java/org/jboss/tools/intellij/openshift/actions/binding/DeleteBindingAction.java [hunk 3]
```diff
@@ -26,6 +26,9 @@ public DeleteBindingAction() {
 
     @Override
     public void actionPerformed(AnActionEvent anActionEvent, Object selected, Odo odo) {
+        if (odo == null) {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/binding/DeleteBindingAction.java:29]: not needed as already taken care of in OdoAction

File: src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/AboutAction.java [hunk 4]
```diff
@@ -32,6 +32,9 @@ public AboutAction() {
 
   @Override
   public void actionPerformed(AnActionEvent anActionEvent, Object selected, Odo odo) {
+    if (odo == null) {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/AboutAction.java:35]: not needed as already taken care of in OdoAction

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsRootNode.java [hunk 5]
```diff
@@ -74,18 +74,21 @@ public void setLogged(boolean logged) {
         this.logged = logged;
     }
 
-    public CompletableFuture<Odo> initializeOdo() {
-        return OdoCliFactory.getInstance().getOdo(project).whenComplete((odo, err) -> {
-            if (this.odo != null) {
-                this.odo.release();
-            }
-            this.odo = new OdoProjectDecorator(odo, this);
-            loadProjectModel(project);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsRootNode.java:83]: this call is still required when loading the root node in applicationTreeStructure class

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java (hunk 2 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java:185]: I believe that I cannot remove `OdoProjectDecorator`. Turns out that this "decorator" does things to the tree before/after invoking the odo binary. Example:
```java
    public void deleteComponent(String project, String context, String component, ComponentKind kind) throws IOException {
        if (context != null) {
            root.removeContext(new File(context));
        }
        delegate.deleteComponent(project, context, component, kind);
    }
```
I could implement the `Odo` interface in `ApplicationsRootNode` but would loose the separation of concerns that we have now. 
To me it looks as if keeping the decorator is the better option. But I'll rename it to `ApplicationsRootNodeOdo` to better reflect its purpose.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java (hunk 1 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoProjectDecorator.java:82]: I decided to keep `OdoProjectDecorator`. See arguments in comments below.

**Comment by adietish** [comment]
@sbouchet corrected what you requested. Please re-review.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java [hunk 6]
```diff
@@ -36,12 +40,32 @@ protected OdoAction(Class... filters) {
     @Override
     public void actionPerformed(AnActionEvent anActionEvent, TreePath path, Object selected) {
         telemetrySender = new TelemetrySender(PREFIX_ACTION + getTelemetryActionName());
-        this.actionPerformed(anActionEvent, (Object) getElement(selected), getOdo(anActionEvent));
+        Odo odo = getOdo(anActionEvent);
+        if (odo == null) {
+          return;
+        }
+        this.actionPerformed(anActionEvent, (Object) getElement(selected), odo);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java:47]: at this point, we're sure that odo isn't null, so maybe add notNull annotation on the abstract might help for docs purpose.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/binding/DeleteBindingAction.java (hunk 3 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/binding/DeleteBindingAction.java:29]: see above.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java (hunk 6 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java:47]: We're not sure that it isn't `null`. See:
```Java
       Odo odo = getOdo(anActionEvent);
        if (odo == null) {
          return;
        }
```
getOdo() can return `null` in many cases:

```
    private Odo getOdo(AnActionEvent anActionEvent) {
        try {
          Tree tree = getTree(anActionEvent);
          if (tree == null) {
            return null;
          }
          ApplicationsTreeStructure structure = (ApplicationsTreeStructure) tree.getClientProperty(Constants.STRUCTURE_PROPERTY);
          if (structure == null) {
            return null;
          }
          ApplicationsRootNode root = (ApplicationsRootNode) structure.getApplicationsRoot();
          if (root == null) {
            return null;
          }
          return root.getOdo().getNow(null);
        } catch(Exception e) {
          LOGGER.warn("Could not get odo: " + e.getMessage(), e);
          return null;
        }
    }
```

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java [hunk 7]
```diff
@@ -123,11 +123,19 @@ private  boolean isDebug(Component component) {
     }
 
     private  boolean isLogRunningForDevOrDebug(ComponentNode componentNode) throws IOException {
-        return componentNode.getRoot().getOdo().isLogRunning(componentNode.getComponent().getPath(), componentNode.getComponent().getName(), false);
+        Odo odo = componentNode.getRoot().getOdo().getNow(null);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java:126]: odo is already present before this call, better to pass it along. and it's also not null at this point ( from OdoAction )

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java [hunk 8]
```diff
@@ -123,11 +123,19 @@ private  boolean isDebug(Component component) {
...
+        if (odo == null) {
+            return false;
+        }
+        return odo.isLogRunning(componentNode.getComponent().getPath(), componentNode.getComponent().getName(), false);
     }
 
     private  boolean isLogRunningForDeploy(ComponentNode componentNode) throws IOException {
-        return componentNode.getRoot().getOdo().isLogRunning(componentNode.getComponent().getPath(), componentNode.getComponent().getName(), true);
+        Odo odo = componentNode.getRoot().getOdo().getNow(null);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java:134]: same as above

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationRootNodeOdo.java [hunk 9]
```diff
@@ -80,14 +90,7 @@ public List<ComponentMetadata> analyze(String path) throws IOException {
...
         if (StringUtils.isNotBlank(starter)) {
-            File tmpdir;
-            if (SystemInfo.isWindows) {
-                tmpdir = Files.createTempDirectory("odotmp").toFile();
-            } else {
-                FileAttribute<Set<PosixFilePermission>> attr = PosixFilePermissions.asFileAttribute(
-                        PosixFilePermissions.fromString("rwxr-x---"));
-                tmpdir = Files.createTempDirectory("odotmp", attr).toFile();
-            }
+            File tmpdir = FileUtil.createTempDirectory("odotmp", null);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationRootNodeOdo.java:93]: good catch of using this IDE util :+1:

**Review by sbouchet — COMMENTED** [review]
please also review the sonar code smells

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java (hunk 7 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java:126]: can do, I agree. But these methods are also called from #isVisible where odo isnt' present. So we need to add odo retrieval there.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/project/CreateProjectAction.java [hunk 10]
```diff
@@ -50,14 +51,18 @@ private void doActioPerformed(ApplicationsRootNode clusterNode) {
     if ((projectName != null) && projectName.trim().length() > 0) {
       CompletableFuture.runAsync(() -> {
         try {
+          Odo odo = clusterNode.getOdo().getNow(null);
```
[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/project/CreateProjectAction.java:54]: Can do, I agree. But then we also need to add odo retrieval in #execute where #doActionPerformed is called, too

**Comment by adietish** [comment]
> please also review the sonar code smells

There are none left. Sonar keeps reporting 2 false positives which I need to close manually on each push.

**Comment by adietish** [comment]
@sbouchet corrected things according to your comments. Please re-review.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java (hunk 6 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/OdoAction.java:47]: maybe i'm wrong but 
```
        if (odo == null) {
          return;
        }
```
will break execution if `odo` is null, meaning that  on next line,

`this.actionPerformed(anActionEvent, (Object) getElement(selected), odo);`

odo isn't null for the rest of the execution ?

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java (hunk 7 above)
Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java]: ok then, better to pass a not null odo to those methods.

File: src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsTreeStructure.java [hunk 11]
```diff
@@ -79,12 +77,6 @@ public ApplicationsTreeStructure(Project project) {
     }
 
     public Object getApplicationsRoot() {
-        if (!initialized.getAndSet(true)) {
```
[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/tree/application/ApplicationsTreeStructure.java:82]: I believe that this call isn't necessary any more. All calls to `getApplicationsRoot()` then get odo from it. Odo then would call `loadProjectModel` if odo isn't initialized yet.

File: src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java (hunk 7 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/component/ShowLogComponentAction.java]: added the annotation on `OdoAction.actionPerformed` and all it's subclasses:

```java
public actionPerformed(AnActionEvent anActionEvent, Object selected, @NotNull Odo odo)
```

**Comment by adietish** [comment]
> seems also that test coverage is missing. need to address that in another issue

I agree. Looked through my changes and all what could make sense here imho is to add a test for `ApplicationsRootNodeOdo` (added/moved in this PR) and file an issue to improve test coverage.

**Comment by adietish** [comment]
@sbouchet I added a unit test for `ApplicationsRootNodeOdo`: https://github.com/redhat-developer/intellij-openshift-connector/pull/589/files#diff-148812f161fd7910c4d4c92503a07efee561cc3fc3ce795f7619aa49db4dc0fa
Please re-review. Thanks.

**Review by sbouchet — APPROVED** [review]
LGTM with finals changes.

**Pull request merged**